src/scenic/core/regions.py

```diff
@@ -281,883 +306,2816 @@ def __hash__(self):
+
+        # Generic case for arbitrary shapes
+        if self.mesh.is_watertight:
+            projection = trimesh.path.polygons.projected(self.mesh, normal=(0,0,1), rpad=1e-4)
+        else:
+            # Special parameters to use all faces if mesh is not watertight.
+            projection = trimesh.path.polygons.projected(self.mesh, normal=(0,0,1), rpad=1e-4,
+                                                         ignore_sign=False, tol_dot=-float('inf'))
+
+        return projection
+
+    @cached_property
+    def boundingPolygon(self):
+        """A PolygonalRegion bounding the mesh"""
+        return PolygonalRegion(polygon=self._boundingPolygon)
+
+    def sampleGiven(self, value):
+        if isinstance(self, MeshVolumeRegion):
+            cls = MeshVolumeRegion
+        elif isinstance(self, MeshSurfaceRegion):
+            cls = MeshSurfaceRegion
+        else:
+            assert False
+
+        return cls(mesh=value[self._mesh], name=self.name, \
+            dimensions=value[self.dimensions], position=value[self.position], rotation=value[self.rotation], \
+            orientation=self.orientation, tolerance=self.tolerance, \
+            centerMesh=self.centerMesh, onDirection=self.onDirection, engine=self.engine)
+
+class MeshVolumeRegion(MeshRegion):
+    """ An instance of MeshRegion that performs operations over the volume of the mesh.
+
+    At intialization time this region assigns the property ``num_samples``,
+    which is the number of samples to attempt to get at least a 99% probability
+    of one being in the volume. This can be overwritten after initialization time,
+    but if it is too low a `RejectionException` is likely and if it is too high sampling
+    can be prolonged.
+    """
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        if needsSampling(self):
+            return
+
+        # Ensure the mesh is watertight so volume is well defined
+        if not self._mesh.is_volume:
+            raise ValueError("A MeshVolumeRegion cannot be defined with a mesh that does not have a well defined volume.")
+
+        # Compute how many samples are necessary to achieve 99% probability
+        # of success when rejection sampling volume.
+        p_volume = self._mesh.volume/self._mesh.bounding_box.volume
+
+        if p_volume > 0.99:
+            self.num_samples = 1
+        else:
+            self.num_samples = min(1e6, max(1, math.ceil(math.log(0.01, 1 - p_volume))))
+
+    # Property testing methods #
+    def intersects(self, other, triedReversed=False):
+        """Check if this region intersects another.
+
+        This function handles intersect calculations for `MeshVolumeRegion` with:
+        * `MeshVolumeRegion`
+        * `MeshSurfaceRegion`
+        * `PolygonalFootprintRegion`
+        """
+        # Check if region is fixed, and if not returns a default implementation
+        if needsSampling(self):
+            return super().intersects(other)
```

Comment:
Needs to pass `triedReversed`. (Better add a test for intersecting two random mesh regions: I think this would enter infinite recursion currently!)